### PR TITLE
config: automatically discover tidb.pd-addr and tidb.port if not provided

### PR DESCRIPTION
**Pull request by kennytm**
### What problem does this PR solve? 

Part of #131. 

The setting `tidb.port` and `tidb.pd-addr` can be automatically discovered through the status API, allowing these to be omitted while Lightning could still obtain the correct values.

### What is changed and how it works?

Change the default values of `tidb.port` from 4000 to 0 and `tidb.pd-addr` from "127.0.0.1:2379" to "".

If these values are empty, reach for `http://{tidb.host}:{tidb.status-port}/settings`, and use these settings instead.

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to update the documentation

### CODE REVIEW

File: lightning/config/config.go [hunk 1]
```diff
@@ -286,9 +290,47 @@ func (cfg *Config) Load() error {
...
+				}
+			}
+			if cfg.TiDB.Port <= 0 {
+				cfg.TiDB.Port = settings.Port
+			}
+		}
+		if len(cfg.TiDB.PdAddr) == 0 {
+			pdAddrs := strings.Split(settings.Path, ",")
+			cfg.TiDB.PdAddr = pdAddrs[0] // FIXME support multiple PDs once importer can.
+		}
```
Comment (GregoryIan) [inline lightning/config/config.go]: check `cfg.TiDB.Port` and `cfg.TiDB.PdAddr` again, if they are empty, throw an error

File: lightning/config/config.go [hunk 2]
```diff
@@ -286,9 +290,47 @@ func (cfg *Config) Load() error {
...
+	if cfg.TiDB.Port <= 0 || len(cfg.TiDB.PdAddr) == 0 {
+		resp, err := http.Get(fmt.Sprintf("http://%s:%d/settings", cfg.TiDB.Host, cfg.TiDB.StatusPort))
+		if err != nil {
+			return errors.Annotate(err, "cannot fetch settings from TiDB, please manually fill in `tidb.port` and `tidb.pd-addr`")
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode != http.StatusOK {
+			return errors.Errorf("TiDB settings returned %s, please manually fill in `tidb.port` and `tidb.pd-addr`", resp.Status)
+		}
+		var settings struct {
```
Comment (GregoryIan) [inline lightning/config/config.go]: use `tidb/config.Config` directly, even it contains too many other fields whose we don't need

File: lightning/config/config.go [hunk 3]
```diff
@@ -286,9 +290,47 @@ func (cfg *Config) Load() error {
...
+		}
+		if cfg.TiDB.Port <= 0 {
+			_, portStr, err := net.SplitHostPort(settings.AdvertiseAddress)
+			if err == nil {
+				port, err := strconv.Atoi(portStr)
+				if err == nil {
+					cfg.TiDB.Port = port
+				}
+			}
+			if cfg.TiDB.Port <= 0 {
```
Comment (GregoryIan) [inline lightning/config/config.go]: put it before L317?

**Comment by GregoryIan** [comment]
Rest LGTM

File: lightning/config/config.go (hunk 3 above)
Comment (kennytm) [reply lightning/config/config.go]: @GregoryIan but `cfg.TiDB.Port` may be modified by line 321

Comment (GregoryIan) [reply lightning/config/config.go]: in fact I mean priority of `settings.Port` is higher than `settings.AdvertiseAddress`

Comment (kennytm) [reply lightning/config/config.go]: If `settings.AdvertiseAddress` has provided a valid port, the port would be set to that and `settings.Port` is ignored as `cfg.TiDB.Port > 0`. So instead `settings.AdvertiseAddress` has higher priority than `settings.Port`. This has been checked in the unit test.

Comment (GregoryIan) [reply lightning/config/config.go]: I ask tidb guys, they told me to use `Port`, it seems `AdvertiseAddress` just is used for PD 😖

Comment (kennytm) [reply lightning/config/config.go]: So it is not supported to put a port-forwarder  in front of TiDB?

Comment (GregoryIan) [reply lightning/config/config.go]: According to the information I understand, it is indeed

**Comment by GregoryIan** [comment]
LGTM

**Comment by lonng** [comment]
@kennytm Please resolve the conflicts

**Comment by kennytm** [comment]
PTAL @lonng

**Pull request merged**